### DISCUSSION TITLE
Can't access property "removeEventListener" when destroy is called before swiper is mounted

### DISCUSSION THREAD

**Discussion by nWacky** [nolimits4web/swiper#6721] [Q&A] [answered]
Hello

I'm using swiper in nuxt. Sometimes I was getting the exception*:

> Error: "TypeError: can't access property "removeEventListener", el[domMethod] is undefined"

The issue was that, when being initialized, swiper allows to be [initialized for a class that doesn't exist yet](https://github.com/nolimits4web/swiper/blob/db7e72f8f3b67049395302d25f2548062f094d54/src/core/core.js#L71-L80) in DOM without an exception. Then swiper can be mounted onto the element later.

When destroying, it only [checks if swiper instance exists](https://github.com/nolimits4web/swiper/blob/db7e72f8f3b67049395302d25f2548062f094d54/src/vue/swiper.js#L348-L350) and [isn't destroyed](https://github.com/nolimits4web/swiper/blob/db7e72f8f3b67049395302d25f2548062f094d54/src/core/core.js#L610). 

If swiper is never mounted, `el` is [a string containing a class name or undefined](https://github.com/nolimits4web/swiper/blob/db7e72f8f3b67049395302d25f2548062f094d54/src/core/core.js#L478C1-L481). In this case destroy throws when unmounting events because [el['removeEventListener'] is undefined](https://github.com/nolimits4web/swiper/blob/db7e72f8f3b67049395302d25f2548062f094d54/src/core/events/index.js#L22).

I'm not sure if
- this should be caught inside a vue component (perhaps, in `swiper/vue`, try/catch destroy [inside onBeforeUnmount](https://github.com/nolimits4web/swiper/blob/db7e72f8f3b67049395302d25f2548062f094d54/src/vue/swiper.js#LL349C9-L349C9)) 
- or if `swiper/core` should handle this case, as `el` can be a string before swiper is mounted.

There's a similar issue [#4043](https://github.com/nolimits4web/swiper/issues/4043), but it didn't say anything conclusive

\* the exception happened because back button was clicked twice with suspense and for some reason nuxt didn't execute `onMounted` hook, but did call `onBeforeUnmount`, I couldn't create a reproduction. Probably [similar issue](https://github.com/nuxt/nuxt/issues/13879)

---

[codesandbox reproduction example](https://codesandbox.io/p/sandbox/swiper-typerror-qxcm09?file=%2Findex.html%3A1%2C1)

**when html element exists**

1. initialize swiper
2. swiper works
3. destroy swiper
4. no exceptions occur

**if element is deleted**

1. initialize swiper
2. no exceptions occur
3. destroy swiper
4. TypeError: can't access property "removeEventListener"

**Comment by tusharbansode** [comment, marked as answer]
export default defineNuxtPlugin((nuxtApp) => {
    
    const testSwiperError = ( error, instance, info ) => {
        return (
            error instanceof TypeError &&
            error.message?.includes("Cannot read properties of undefined (reading 'removeEventListener')") &&
            (error.stack?.includes("swiper-core") || error.stack?.includes("swiper_vue") || error.stack?.includes("swiper/modules"))
        );
    }

    nuxtApp.vueApp.config.errorHandler = (error, instance, info) => {
      
        if (testSwiperError(error, instance, info)) {
            console.warn("[Handled Swiper Error]:", error.message);
            return;
        }
        throw error;
    }
  
    nuxtApp.hook('vue:error', (error, instance, info) => {
        
        if (testSwiperError(error, instance, info)) {

            console.warn("[Handled Swiper Error]:", error.message);
            return;
        }
        
        throw error;
    })
})

**Comment by nWacky** [comment]
Yes. 

It's been a while since I asked this question

[NuxtErrorBoundary](https://nuxt.com/docs/api/components/nuxt-error-boundary) can also be used to catch any swiper errors

```vue
<script setup lang="ts">
const logSomeError = (e: unknown) => {
  console.error("Swiper error", e);
};
</script>

<template>
  <NuxtErrorBoundary @error="logSomeError">
    <swiper
      :slides-per-view="3"
      :space-between="50"
    >
      <swiper-slide>Slide 1</swiper-slide>
      <swiper-slide>Slide 2</swiper-slide>
      <swiper-slide>Slide 3</swiper-slide>
    </swiper>
  </NuxtErrorBoundary>
</template>
```
